src/bro_detective/from_files.rs

```diff
@@ -0,0 +1,60 @@
+use std::collections::HashMap;
+use regex::Regex;
+use std::fs;
+
+extern crate yaml_rust;
+use yaml_rust::YamlLoader;
+
+
+pub fn detect_potential_makers(detectors: &HashMap<String, Vec<String>>, project_dir: String) -> Result<Vec<String>, String> {
+    let mut results = Vec::new();
+    let project_contents = fs::read_dir(project_dir).unwrap();
+    let mut project_root_filenames = String::new();
+    for path in project_contents {
+        let path = path.unwrap().file_name();
+        let path = path.to_str().unwrap();
```

Comment:
A lot of `unwrap`s here, since the function returns a result it shouldn't be too hard to convert everything to it and handle the errors gracefully using the question mark operator:
```rust
let project_contents = fs::read_dir(project_dir).map_err(|e| format!("error reading directory: {e}"))?;
```

Another thing you can do here is use the `std::io::Error`, since you most of your errors are actually IO errors including this one, it might make sense to return an `io::Error` everywhere.
If you want to create an `io::Error` by yourself and none of the `io:Error` variants feel right, you can use the [`std::io::Error::other`](https://doc.rust-lang.org/std/io/struct.Error.html#method.other) to easily create your own custom IO error:
```rust
pub fn detect_potential_makers(/* ... */) -> Result<Vec<String>, io::Error> {
   // ...
   // the `?` operator "just works", because the errors are of the same type!
   let project_contents = fs::read_dir(project_dir)?;
   // ...
       // ...
       // the error from Regex is not an io::Error, but we can easily make it into one by using 'other':
       let re = Regex::new(&detect_file).map_err(|e| io::Error::other(e.to_string()))?;
}
```

When the project gets bigger it is common to create your own error enum, and to avoid using `map_err` everywhere you usually implement your own conversion methods from other errors to yours (the `From` trait)

Useful crates for error handling: `anyhow`, `thiserror`

Reply:
Great this is what I'm looking for :pray:

Reply:
Refactored fs stuff completely.